### ISSUE TITLE
Compiling on OS X

### ISSUE THREAD

**Issue by camspiers** [preillyme/v8js#97]
I have had a lot of trouble compiling on OS X using the method described in the readme.

I have got it working with the following instead (requires g++-4.7).

```
cd /tmp
git clone https://github.com/v8/v8.git
cd v8
make dependencies

export CXX=/usr/local/bin/g++-4.7
export LINK=/usr/local/bin/g++-4.7
make native component=shared_library snapshot=off -j8

# Important to turn snapshot off

sudo mkdir -p /tmp/v8-install/lib /tmp/v8-install/include

# Copy the dylib as an "so" doesn't get built
sudo cp out/native/lib*.dylib /tmp/v8-install/lib/
sudo cp include/v8* /tmp/v8-install/include

cd /tmp
git clone https://github.com/preillyme/v8js.git
cd v8js
phpize

# Need to configure with D_GLIBCXX_USE_NANOSLEEP
./configure --with-v8js=/tmp/v8-install CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"
make
```

I also needed to remove the "-Wl,-rpath" flags in the config.m4 file:

```
dnl $Id$

PHP_ARG_WITH(v8js, for V8 Javascript Engine,
[  --with-v8js           Include V8 JavaScript Engine])

if test "$PHP_V8JS" != "no"; then
  SEARCH_PATH="/usr/local /usr"
  SEARCH_FOR="include/v8.h"
  
  if test -r $PHP_V8JS/$SEARCH_FOR; then
    LDFLAGS="$LDFLAGS"
    V8_DIR=$PHP_V8JS
  else
    AC_MSG_CHECKING([for V8 files in default path])
    for i in $SEARCH_PATH ; do
      if test -r $i/$SEARCH_FOR; then
        V8_DIR=$i
        AC_MSG_RESULT(found in $i)
      fi
    done
  fi

  if test -z "$V8_DIR"; then
    AC_MSG_RESULT([not found])
    AC_MSG_ERROR([Please reinstall the v8 distribution])
  fi

  PHP_ADD_INCLUDE($V8_DIR/include)
  PHP_ADD_LIBRARY_WITH_PATH(v8, $V8_DIR/$PHP_LIBDIR, V8JS_SHARED_LIBADD)
  PHP_SUBST(V8JS_SHARED_LIBADD)
  PHP_REQUIRE_CXX()

  old_LIBS=$LIBS
  old_LDFLAGS=$LDFLAGS
  old_CPPFLAGS=$CPPFLAGS
  LDFLAGS="-L$V8_DIR/$PHP_LIBDIR"
  LIBS=-lv8
  CPPFLAGS=-I$V8_DIR/include
  AC_LANG_SAVE
  AC_LANG_CPLUSPLUS

  AC_CACHE_CHECK(for V8 version, ac_cv_v8_version, [
AC_TRY_RUN([#include <v8.h>
#include <iostream>
#include <fstream>
using namespace std;

int main ()
{
	ofstream testfile ("conftestval");
	if (testfile.is_open()) {
		testfile << v8::V8::GetVersion();
		testfile << "\n";
		testfile.close();
		return 0;
	}
	return 1;
}], [ac_cv_v8_version=`cat ./conftestval|awk '{print $1}'`], [ac_cv_v8_version=NONE], [ac_cv_v8_version=NONE])
])

  if test "$ac_cv_v8_version" != "NONE"; then
    ac_IFS=$IFS
    IFS=.
    set $ac_cv_v8_version
    IFS=$ac_IFS
    V8_API_VERSION=`expr [$]1 \* 1000000 + [$]2 \* 1000 + [$]3`
    if test "$V8_API_VERSION" -lt 3021012 ; then
       AC_MSG_ERROR([libv8 must be version 3.21.12 or greater])
    fi
    AC_DEFINE_UNQUOTED([PHP_V8_API_VERSION], $V8_API_VERSION, [ ])
    AC_DEFINE_UNQUOTED([PHP_V8_VERSION], "$ac_cv_v8_version", [ ])
  else
    AC_MSG_ERROR([could not determine libv8 version])
  fi

  AC_CACHE_CHECK(for debuggersupport in v8, ac_cv_v8_debuggersupport, [
AC_LINK_IFELSE([AC_LANG_PROGRAM([[#include <v8-debug.h>]],
                               [[v8::Debug::DisableAgent()]])],
    [ac_cv_v8_debuggersupport=yes],
    [ac_cv_v8_debuggersupport=no])
])

  if test "$ac_cv_v8_debuggersupport" = "yes"; then
    AC_DEFINE([ENABLE_DEBUGGER_SUPPORT], [1], [Enable debugger support in V8Js])
  fi

  AC_LANG_RESTORE
  LIBS=$old_LIBS
  LDFLAGS=$old_LDFLAGS
  CPPFLAGS=$old_CPPFLAGS


  AC_CACHE_CHECK(for C standard version, ac_cv_v8_cstd, [
    ac_cv_v8_cstd="c++11"
    old_CPPFLAGS=$CPPFLAGS
    AC_LANG_PUSH([C++])
    CPPFLAGS="-std="$ac_cv_v8_cstd
    AC_TRY_RUN([int main() { return 0; }],[],[ac_cv_v8_cstd="c++0x"],[])
    AC_LANG_POP([C++])
    CPPFLAGS=$old_CPPFLAGS
  ]);
  
  PHP_NEW_EXTENSION(v8js, v8js.cc v8js_convert.cc v8js_methods.cc v8js_variables.cc v8js_commonjs.cc, $ext_shared, , "-std="$ac_cv_v8_cstd)

  PHP_ADD_MAKEFILE_FRAGMENT
fi
```

**Comment by AlexeyFrolov** [comment]
cant build on Maverics too.. (

```
➜  v8js git:(master) pwd
/Users/fro/build/v8js
➜  v8js git:(master) ./configure --with-v8js=/Users/fro/build/v8 CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"
checking for grep that handles long lines and -e... /usr/local/bin/ggrep
checking for egrep... /usr/local/bin/ggrep -E
checking for a sed that does not truncate output... /usr/local/bin/gsed
checking for cc... cc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether cc accepts -g... yes
checking for cc option to accept ISO C89... none needed
checking how to run the C preprocessor... cc -E
checking for icc... no
checking for suncc... no
checking whether cc understands -c and -o together... yes
checking for system library directory... lib
checking if compiler supports -R... no
checking if compiler supports -Wl,-rpath,... yes
checking build system type... x86_64-apple-darwin14.1.0
checking host system type... x86_64-apple-darwin14.1.0
checking target system type... x86_64-apple-darwin14.1.0
checking for PHP prefix... /usr/local/Cellar/php56/5.6.5
checking for PHP includes... -I/usr/local/Cellar/php56/5.6.5/include/php -I/usr/local/Cellar/php56/5.6.5/include/php/main -I/usr/local/Cellar/php56/5.6.5/include/php/TSRM -I/usr/local/Cellar/php56/5.6.5/include/php/Zend -I/usr/local/Cellar/php56/5.6.5/include/php/ext -I/usr/local/Cellar/php56/5.6.5/include/php/ext/date/lib
checking for PHP extension directory... /usr/local/Cellar/php56/5.6.5/lib/php/extensions/debug-non-zts-20131226
checking for PHP installed headers prefix... /usr/local/Cellar/php56/5.6.5/include/php
checking if debug is enabled... yes
checking if zts is enabled... yes
checking for re2c... no
configure: WARNING: You will need re2c 0.13.4 or later if you want to regenerate PHP parsers.
checking for gawk... no
checking for nawk... no
checking for awk... awk
checking if awk is broken... no
checking for V8 Javascript Engine... yes, shared
checking for g++... g++
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking how to run the C++ preprocessor... g++ -E
checking for V8 version... NONE
configure: error: could not determine libv8 version

➜  v8  pwd
/Users/fro/build/v8
➜  v8  tree
.
├── include
│   ├── v8-debug.h
│   ├── v8-platform.h
│   ├── v8-profiler.h
│   ├── v8-testing.h
│   ├── v8-util.h
│   ├── v8.h
│   └── v8config.h
└── lib
    ├── libicui18n.dylib
    ├── libicuuc.dylib
    └── libv8.dylib

2 directories, 10 files
```

Why it "could not determine libv8 version"? Thanks in advance!

**Comment by namjul** [comment]
having the same problem like AlexeyFrolov.

**Comment by vojtech-dobes** [comment]
Also affected, running Yosemite 10.10.2. I have v8 installed using brew, version 3.30.33.16.

**Comment by jaromirnyklicek** [comment]
Same problem here (Yosemite 10.10.2). I have overridden this by adding dirty hack to my config.m4 file (around line 50):
```
if (testfile.is_open()) {
        //testfile << v8::V8::GetVersion();
        testfile << "3.30.33.16";
        testfile << "\n";
        testfile.close();
        return 0;
}
```

then `` ./configure --with-v8js=/tmp/v8-install CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"`` ends correctly and creates Makefile.

But when I call 'make' I get couple of theese errors

```
/usr/local/Cellar/php56/5.6.5/include/php/Zend/zend_API.h:112:53: note: expanded from macro 'ZEND_BEGIN_ARG_INFO'
        ZEND_BEGIN_ARG_INFO_EX(name, 0, ZEND_RETURN_VALUE, -1)
                                                           ^~
/usr/local/Cellar/php56/5.6.5/include/php/Zend/zend_API.h:110:20: note: expanded from macro 'ZEND_BEGIN_ARG_INFO_EX'
                { NULL, 0, NULL, required_num_args, 0, return_reference, 0, 0 },
                                 ^
/tmp/v8js/v8js_class.cc:1039:1: error: constant expression evaluates to -1 which cannot be narrowed to type 'zend_uint' (aka 'unsigned int') [-Wc++11-narrowing]
ZEND_BEGIN_ARG_INFO(arginfo_v8js_getextensions, 0)
```

So I guess there is still something wrong...

**Comment by tahpot** [comment]
I'm also on OSX 10.10.2 and can't build. Tried @jaromirnyklicek approach of editing config.m4, but I still receive:

> configure: error: could not determine libv8 version

**Comment by stesie** [comment]
I've finally fetched a MacOS vagrant image from the internet to give it a try myself; I'm usually on just on GNU/Linux boxes so bear with me...

The rpath stuff doesn't seem to be supported by Darwin, hence I've prepared a patched config.m4 file I'll eventually merge.

I used the Apple LLVM 6.0 compiler (clang); in combination with v8 *3.28.71.19* (latest one from the 3.28 line) *it works* for me.  I also got the error @jaromirnyklicek  mentioned

```
/tmp/v8js/v8js_class.cc:1039:1: error: constant expression evaluates to -1 which cannot be narrowed to type 'zend_uint' (aka 'unsigned int') [-Wc++11-narrowing]
ZEND_BEGIN_ARG_INFO(arginfo_v8js_getextensions, 0)
```
... which is caused by PHP internal macros, which cast -1 to unsigned integer.  gcc accepts that, but clang doesn't.  However you can pass `CXXFLAGS="-Wno-c++11-narrowing"` to configure.

----

v8 versions 3.29.36 or later do **not** currently work.  This is because v8 needs a library named libplatform there, which is available but it uses symbols from v8 library that are not externally visible and hence cause runtime problems.

* v8js.so needs symbol `__ZN2v84base5MutexC1Ev` (via statically linked libplatform); capital U for "undefined"
```
vagrant-osx:v8js vagrant$ nm /usr/local/lib/php/extensions/no-debug-non-zts-20131226/v8js.so | grep __ZN2v84base5MutexC1Ev
                 U __ZN2v84base5MutexC1Ev
```
* libv8.dylib does even define it ("t" for text segment aka code), but it's a lower case letter (which means internal)
```
vagrant-osx:v8js vagrant$ nm /usr/local/lib/libv8.dylib | grep __ZN2v84base5MutexC1Ev
00000000003447de t __ZN2v84base5MutexC1Ev
```

* on a GNU/Linux system this symbol is public (capital T)
```
root@97c2d1abbaf4:/# nm /usr/local/v8/lib/libv8.so  | grep -ie ZN2v84base5MutexC1Ev
0000000000677170 T _ZN2v84base5MutexC1Ev
```
... and hence it's all fine there

----

TL;DR
* v8 versions from the 3.28 line supposed to work fine
* need to dig further regarding how to externalize the above symbol on darwin
* I have a draft README.md file for MacOS around and will merge it with the config.m4 patch soonish

cheers
  ~stesie


by the way, I think you all don't need the `CXXFLAGS="-D_GLIBCXX_USE_NANOSLEEP"` flags, they are just a work around needed for gcc 4.7

**Comment by tahpot** [comment]
Thanks stesie, much appreciated. Looking forward to the patch.

**Comment by vojtech-dobes** [comment]
Awesome :).

**Comment by stesie** [comment]
Ok, I've now merged my changes and I've also fixed the libplatform problem that lead to linking issues with v8 versions 3.29.36 and higher

I've tried the 3.30 version installed by `brew install v8` -> works for me (now).
Version 3.32 should work likewise.

Version 4.x are currently known to be broken.  So better don't use these for the moment :-)

-----

@evandrix looks like you're using the 0.1.5 release version, please use master instead.

@camspiers (and all others here actually): you shouldn't need to disable snapshots (`snapshot=off`); indeed the build stops due to libicu*.dylib not being available from /usr/local/lib.  However you can just copy these from out/native/ already and re-start make. Then it is able to create the snapshot, which is a big plus IMHO (performance wise)

@AlexeyFrolov @namjul I've tested with Yosemite, please try with mavericks again and let me know if it still doesn't work.  Thanks!

cheers
  ~stesie

**Comment by jaromirnyklicek** [comment]
Thanks, works great. You're awesome ;-)

**Comment by camspiers** [comment]
@stesie Thanks so much for your hard work on this!

**Comment by tahpot** [comment]
I'm still having problems when compiling v8js:

/tmp/v8js/v8js_class.cc:293:6: error: expected a class or namespace
        v8::Debug::ProcessDebugMessages();
        ~~~~^
/tmp/v8js/v8js_class.cc:635:7: error: expected a class or namespace
                v8::Debug::DisableAgent();
                ~~~~^
/tmp/v8js/v8js_class.cc:672:6: error: expected a class or namespace
        v8::Debug::SetDebugMessageDispatchHandler(DispatchDebugMessages, true);
        ~~~~^
/tmp/v8js/v8js_class.cc:673:6: error: expected a class or namespace
        v8::Debug::EnableAgent(str_len ? str : "V8Js", port, auto_break > 0);
        ~~~~^
/tmp/v8js/v8js_class.cc:679:7: error: expected a class or namespace
                v8::Debug::DebugBreak(c->isolate);

I tried with "brew install v8" and then tried compiling as per the MacOS.md, both with the same results :(

**Comment by tahpot** [comment]
I've got this working once I upgraded v8 from 3.25.30 to the latest (3.30.33.16) after doing "brew update".

Thanks for your excellent work @stesie

**Comment by stesie** [comment]
@tahpot 3.25.30 should work again as well, just in case should you still care after upgrading :)

**Comment by TimNZ** [comment]
Sorry, I'm a noob and frustratingly nothing is working to result in a successful compile.

Followed your README.mac_os instructions, following the guidance here, still get "checking for V8 version... NONE".

brew install v8 is installing v4.x which I did an unlink of and then followed your instructions.
Don't know how to read configure.sh so don't know exactly why it is failing.

I have no idea if I can do a brew install of an earlier version of v8, brew versions doesn't work, and none of the instructions I found worked.

It's like something broke in the last month and now nothing works.

No wonder people use Windows :)

**Comment by TimNZ** [comment]
If this helps:

From config.m4:
```
AC_TRY_RUN([#include <v8.h>
#include <iostream>
#include <fstream>
using namespace std;

int main ()
{
	ofstream testfile ("conftestval");
	if (testfile.is_open()) {
		testfile << v8::V8::GetVersion();
		testfile << "\n";
		testfile.close();
		return 0;
	}
	return 1;
}], [ac_cv_v8_version=`cat ./conftestval|awk '{print $1}'`], [ac_cv_v8_version=NONE], [ac_cv_v8_version=NONE])
```

which ends up in configure after phpize:
```
if test "$cross_compiling" = yes; then :
  ac_cv_v8_version=NONE
else
  cat confdefs.h - <<_ACEOF >conftest.$ac_ext
/* end confdefs.h.  */
#include <v8.h>
#include <iostream>
#include <fstream>
using namespace std;

int main ()
{
	ofstream testfile ("conftestval");
	if (testfile.is_open()) {
		testfile << v8::V8::GetVersion();
		testfile << "\n";
		testfile.close();
		return 0;
	}
	return 1;
}
_ACEOF
if ac_fn_cxx_try_run "$LINENO"; then :
  ac_cv_v8_version=`cat ./conftestval|awk '{print $1}'`
else
  ac_cv_v8_version=NONE
fi
```

The ac_fn_cxx_try fails and the else block is executed => NONE
